### DISCUSSION TITLE
pyodide: additional modules, dynamic loading, administrative constraints, performance

### DISCUSSION THREAD

**Discussion by dagbdagb** [open-webui/open-webui#7312] [Ideas]
I  recently discovered that OWUI allows for executing python code in the browser. 

I find this very useful for learning and perhaps even for teaching math (and python). A model can be asked to explain a math concept/term/method/etc., and then also be asked to visualize the same with python code (which renders an inline illustration directly in the web UI). This is of course a very narrow use-case, but it is the one on my mind at this time.

Executing python code in the browser ensures that abuse doesn't impact anyone else using OWUI or the LLM backend in question. So far so good.

The currently available selection of python modules (in pyodide, as implemented in OWUI) outside of core is somewhat limited:
        'micropip',    (allows for installing pure python packages from pypi*)
        'packaging',
        'requests',
        'beautifulsoup4',
        'numpy',
        'pandas',
        'matplotlib',
        'scikit-learn',
        'scipy',
        'regex',
        'seaborn',  (not a pyodide 'native' package)

For math in particular, the following python modules are useful, highly relevant and do exist as native pyodide packages:
        'sympy',
        'geopandas',
        'statsmodels',
        'networkx',
        'bokeh',
        'altair',

*) micropip permits installing pure python packages from pypi, which may fail for a variety of reasons. 
pyodide.loadPackage only installs 'native' pyodide packages and resolves dependencies correctly.
See [discussion](https://github.com/pyodide/pyodide/issues/870).

"Any pure Python package with a wheel available on PyPI is supported. Many packages with C extensions have also been ported for use with Pyodide. These include many general-purpose packages such as regex, pyyaml, lxml and scientific Python packages including numpy, pandas, scipy, matplotlib, and scikit-learn."

What I do not know is if and how adding additional modules may affect the user experience in general. I assume there is a reason the number of modules were somewhat limited at the outset. But I don't know for sure.

Also, naming and adding each and every explicitly permitted module in the source code of OWUI appears cumbersome. Is it plausible that a mechanism can be devised which permits the OWUI administrator to check the individual modules the users can make use of? And if so, would this actually provide value in terms of lowered resource use in the browser, or in the front- or  backend?

Would such a mechanism be more useful if it permits the user to micropip.install from an administrator-approved whitelist of pure python modules? (Assuming (....) all native pyodide modules are safe in the context of OWUI.)

I expect I have managed to convey my 'green-ness' w.r.t. both OWUI and pyodide. I don't know what I don't know. But I am fairly certain there are things I don't know or understand yet.

Also, I'd like to thank @mafzaal for publishing a sample patch I could crib from to test a few additional modules in OWUI.

**Comment by peter-ch** [comment]
I tried everything to add sympy and nothing worked. Can someone tell me how to do this, it can't possibly be hard?

**Reply by dagbdagb** [reply]
I find it slightly cumbersome, actually. I managed to add a nice handful of packages by cribbing from the work done by @mafzaal :
https://github.com/mafzaal/open-webui/commit/d6a2e0fe9bff0d87c108f9ecdeb6848d6e782f4a

+1 for sympy. Very, very useful for math.

I intended to cook a patch for adding in sympy and a few other packages, but got sidetracked and ran out of steam.
